src/battle_ai_switch_items.c

```diff
@@ -208,28 +231,81 @@ static bool32 HasBadOdds(u32 battler, bool32 emitResult)
 
             // Switch mon out
             gBattleStruct->AI_monToSwitchIntoId[battler] = PARTY_SIZE;
-            if (emitResult)
-                BtlController_EmitTwoReturnValues(battler, 1, B_ACTION_SWITCH, 0);
-            return TRUE;
+            if (IsSwitchinIdValid(battler, gBattleStruct->AI_monToSwitchIntoId[battler]))
+            {
+                if (emitResult)
+                    BtlController_EmitTwoReturnValues(battler, 1, B_ACTION_SWITCH, 0);
+                return TRUE;
+            }
         }
     }
     return FALSE;
 }
 
-static bool32 ShouldSwitchIfAllBadMoves(u32 battler, bool32 emitResult)
+static bool32 ShouldSwitchIfAllMovesBad(u32 battler, bool32 emitResult)
 {
-    if (AI_DATA->shouldSwitchIfBadMoves & (1u << battler))
+    u32 battlerIndex, moveIndex;
+    bool32 switchOut = FALSE;
+
+    // Consider switching if all moves are worthless to use.
+    if (GetTotalBaseStat(gBattleMons[battler].species) >= 310 // Mon is not weak.
+        && gBattleMons[battler].hp >= gBattleMons[battler].maxHP / 2) // Mon has more than 50% of its HP
     {
-        AI_DATA->shouldSwitchIfBadMoves &= ~(1u << battler);
-        gBattleStruct->AI_monToSwitchIntoId[battler] = AI_DATA->monToSwitchId[battler];
-        if (emitResult)
-            BtlController_EmitTwoReturnValues(battler, BUFFER_B, B_ACTION_SWITCH, 0);
-        return TRUE;
+        s32 cap = AI_THINKING_STRUCT->aiFlags[battler] & (AI_FLAG_CHECK_VIABILITY) ? 95 : 93;
+        if (IsDoubleBattle())
+        {
+            for (battlerIndex = 0; battlerIndex < MAX_BATTLERS_COUNT; battlerIndex++)
+            {
+                if (battlerIndex != battler && IsBattlerAlive(battlerIndex))
+                {
+                    for (moveIndex = 0; moveIndex < MAX_MON_MOVES; moveIndex++)
+                    {
+                        if (gBattleStruct->aiFinalScore[battler][battlerIndex][moveIndex] > cap)
+                            break;
+                    }
+                    if (moveIndex != MAX_MON_MOVES)
+                        break;
+                }
+            }
+            if (battlerIndex == MAX_BATTLERS_COUNT && AI_DATA->mostSuitableMonId[battler] != PARTY_SIZE)
+                switchOut = TRUE;
+        }
+        else
+        {
+            for (battlerIndex = 0; battlerIndex < MAX_MON_MOVES; battlerIndex++)
```

Comment:
shouldn't this be using `moveIndex`? Doesn't matter for code functionality but still

Reply:
Hilariously it wasn't in the original code it was using i which is battlerIndex. I left it because it was there originally but that's a good point.

I think this is actually the opposite where MAX_MON_MOVES is meant to be MAX_BATTLERS_COUNT rather than the index being wrong. Have pushed this change also :)